### DISCUSSION TITLE
Which code handles file: protocol?

### DISCUSSION THREAD

**Discussion by guest271314** [curl/curl#13234] [Q&A] [answered]
What I'm trying to do is extract only those parts to use as a standalone executable and/or as souce for WASM. Then that code will be used to implement fetching `file:` protocol in Node.js. I built `curl` (on Linux) and this works, I just don't need all the rest of what `curl` does in the executable. I also noticed that when building `curl` still expects `include` folder to be present; if not will throw can't find symbol error.

What parts of `curl` handle fetching `file:` protocol?

**Comment by bagder** [comment, marked as answer]
> I also noticed that when building curl still expects include folder to be present; if not will throw can't find symbol error.

Yes that is correct. It needs the headers.

> What parts of curl handle fetching file: protocol?

Where did you look? it is in the what I would consider rather cleverly named file `lib/file.c`...

**Reply by guest271314** [reply]
I read https://curl.se/docs/install.html. I just began exploring `curl` as a potential solution to what I am trying to do (implement support for `file:` URL's in Node.js' `fetch()`) yesterday, see https://gist.github.com/guest271314/a4f005d9a6b5b433ae6d6e6c5c6d7595. 

This is what I have so far, that works

```
// node --experimental-default-type=module fetchFile.js
/*
git clone https://github.com/curl/curl.git
cd curl
autoreconf -fi
LDFLAGS="-static" ./configure --disable-alt-svc --disable-ares --disable-cookies --disable-basic-auth --disable-bearer-auth --disable-digest-auth --disable-kerberos-auth --disable-negotiate-auth --disable-aws --disable-dateparse --disable-dnsshuffle --disable-doh --disable-form-api --disable-get-easy-options --disable-hsts --disable-http-auth --disable-ipv6 --disable-libcurl-option --disable-manual --disable-ntlm --disable-ntlm-wb --disable-progress-meter --disable-proxy --disable-pthreads  --disable-socketpair --disable-threaded-resolver --disable-tls-srp --disable-unix-sockets --disable-versioned-symbols --without-brotli --without-libpsl --without-nghttp2 --without-ngtcp2 --without-zstd --without-libidn2 --without-librtmp --without-ssl --without-zlib --enable-static --prefix=$HOME/bin
make -j $(nproc)
make install
mv bin/* $PWD
rm -rf bin lib include share
*/
import { spawn } from "node:child_process";
import { Duplex } from "node:stream";

async function fetchFile(path) {
  // https://github.com/chcunningham/atomics-post-message/blob/main/server.js
  const mimeTypes = {
    "html": "text/html",
    "jpeg": "image/jpeg",
    "jpg": "image/jpeg",
    "png": "image/png",
    "js": "text/javascript",
    "wasm": "application/wasm",
    "css": "text/css",
  };
  try {
    const { stdout, stderr } = spawn("./bin/curl", ["-NSs", path]);
    if (stderr) {
      const err = await new Response(Duplex.toWeb(stderr).readable).text();
      if (err) {
        throw err;
      }
    }
    return new Response(Duplex.toWeb(stdout).readable, {
      headers: {
        "Content-Type": mimeTypes[path.split(".").pop()] || "text/plain",
      },
    });
  } catch (e) {
    throw e;
  }
}

export { fetchFile };
```

The reason I asked this question is after building `curl` I tried to move `curl` to the parent directory and delete the curl repository. I can do that when building txiki.js and QuickJS. 

Is there any way to compile `curl` to a standalone executable and/or as a WASM module, e.g., using WASI?

**Reply by guest271314** [reply]
Should I ask another question about how to compile `lib/file.c` to a standalone executable? Or can you kindly answer that inquiry here?

**Reply by bagder** [reply]
That's not possible. libcurl is a transfer library that supports many different protocols, `lib/file.c` features the file: specific code but it also needs the core library logic in order to work.

**Reply by guest271314** [reply]
Thanks.  I see somebody has compiled `curl` to WASM https://github.com/wapm-packages/curl. I'm just trying to use the least amount of working code to achieve my own requirement simply to do it within the restrictions of a random StackOverflow question.

**Comment by dfandrich** [comment]
Keep in mind that just keeping the file handler around isn't nearly enough.
There's a lot of support code that's needed for a file: transfer to occur via
libcurl. If that weren't the case, you would just be using the POSIX file API
instead.

Rather than trying to extract the needed code manually, you'll have quicker
success by just using the build system to enable only the file: protocol and
disable all the others.  It will still compile the other files but many of
them will end up as zero-length and won't contribute to the final size of the
library. See the Reducing Size section of docs/INSTALL.md for tips in using
this approach.

**Reply by guest271314** [reply]
I read the document https://curl.se/docs/install.html. I can build `tjs` from txiki.js and  `qjs` QuickJS, a dependency of txiki.js from the repository, move the standalone executables anywhere on the file system and they work as intended, without the parent GitHub repository. I was thinking I could do that for `curl`.

**Reply by guest271314** [reply]
> you'll have quicker
success by just using the build system to enable only the file: protocol and
disable all the others.

I must have missed 

> Besides specifying the --disable-PROTOCOL flags
